### PR TITLE
S5 megan

### PR DESCRIPTION
**Pull request by meganmansfield**
Finished editing so step 5 (lightcurve fitting) uses a .ecf file to reference the parameters. This still only works on a pure transit model, and the additional systematics you can fit for aren't included yet.

### CODE REVIEW

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:15]: `from . import parameters as p`

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:30]: `from . import lightcurve as lc`

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:18]: This comment doesn't seem related to the code below, and it isn't clear to me that the code below would work in general. Where is the speclc folder made, and what .txt files are supposed to be in it? No such folder or text files are always made as an output from stage 4.

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:24]: Why is this written as a for-loop if you're just taking the first element of files? I guess the interpretation of what you're doing here depends on what the files you're reading in are

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:25]: Oh I see, so this is supposed to be a csv type output from Stage 4? Currently it outputs only a single .dat file

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:35]: `from .utils import get_target_data`

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:40]: If these commented out lines no longer serve any purpose, it'd be best to cut them out to keep files smaller and easier to read through

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (taylorbell57) [inline eureka/S5_lightcurve_fitting/s5_fit.py:58]: `from . import models as m`

File: eureka/lib/readECF.py
Comment (taylorbell57) [inline eureka/lib/readECF.py:101]: If you no longer need these commented out edits you made, best to delete them

File: demos/S5/run_S5.py
Comment (taylorbell57) [reply demos/S5/run_S5.py:20]: It'd be great if Stage 5 could be run on all of the different source and background aperture pairs considered in Stage 3 in the same way as how I wrote Stage 4. That way we can just say explore all these different aperture pairs and let the final fits determine which we want to go with in the end. Again, happy to help with this or you can look at what I did for Stage 4

**Review by taylorbell57 — COMMENTED** [review]
Alright, I gave your PR a read through and left my thoughts throughout - you can reply to the comments in-thread if you want to clarify anything, and pushing any new code to this same branch on GitHub will update your PR.

File: demos/S5/run_S5.py
Comment (meganmansfield) [reply demos/S5/run_S5.py:14]: Will do! What about the other parameters here? Hypothetically eventlabel and meta could also go in an ecf file, but I don't know if that would be in line with what was done in stages 2-4.

File: demos/S5/run_S5.py
Comment (meganmansfield) [reply demos/S5/run_S5.py:12]: Do you mean this functionality already exists (the meta file is already somewhere in there) or we need to add something to S5_lightcurve_fitting that lets them access it without adding the file names manually?

File: demos/S5/run_S5.py
Comment (taylorbell57) [reply demos/S5/run_S5.py:14]: You should check out what is present in Stages 2-4 [here](https://github.com/kevin218/Eureka/blob/11120b215e9c0e0a54de8a5a232874b181b67efe/demos/S2/S2_miri_lrs_template.ecf), [here](https://github.com/kevin218/Eureka/blob/11120b215e9c0e0a54de8a5a232874b181b67efe/demos/S3/S3_template.ecf), and [here](https://github.com/kevin218/Eureka/blob/11120b215e9c0e0a54de8a5a232874b181b67efe/demos/S4/S4_template.ecf). The eventlabel needs to be in the run_eureka.py type file (example [here](https://github.com/kevin218/Eureka/blob/11120b215e9c0e0a54de8a5a232874b181b67efe/demos/S2/run_eureka.py)), but basically everything else should be read in from the ecf.

File: demos/S5/run_S5.py
Comment (taylorbell57) [reply demos/S5/run_S5.py:12]: You'll need to add something to the S5 file that is similar to what is done in S3 and S4

File: demos/S5/s5.ecf
Comment (meganmansfield) [reply demos/S5/s5.ecf:11]: Half and half of each. plot_raw_data, fit_white, divide_white are ones that I wasn't sure if we would need them (any thoughts?) The limb darkening parameters (fix_ld and ld_file) are things that I imagine we might want at some point but I believe aren't functional yet. For the sigma clipping parameters (clipsigma and clipiters), I wasn't sure if this was something we needed for S5 or if sigma clipping had already happened somewhere else in the pipeline. So a mix of things that might be added at some point and things that I wasn't sure whether they are obsolete. If you have opinions, I can edit to match whatever you think would work best!

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (meganmansfield) [reply eureka/S5_lightcurve_fitting/fitters.py:742]: Ok, I'll check that out and add it.

File: demos/S5/run_S5.py
Comment (meganmansfield) [reply demos/S5/run_S5.py:20]: I can take a look at what you wrote for Stage 4, or if you have time and want to try out connecting them, that would be great!

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (meganmansfield) [reply eureka/S5_lightcurve_fitting/s5_fit.py:18]: This code is what Sebastian set up to run on the lightcurve files we were using before we had output from the full pipeline. So basically right now Stage 5 is disconnected from all the others and just works on this directory of files Sebastian set up. I think this part could get removed when we connect Stage 5 to Stage 4.

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (meganmansfield) [reply eureka/S5_lightcurve_fitting/s5_fit.py:24]: Yeah, again, this is some code that was only there when Stage 5 was running on files not output from the rest of Eureka.

File: eureka/S5_lightcurve_fitting/s5_fit.py
Comment (meganmansfield) [reply eureka/S5_lightcurve_fitting/s5_fit.py:25]: All these parts were written before I started working on this - I can try and change them to connect to Stage 4 at some point.

File: eureka/lib/readECF.py
Comment (meganmansfield) [reply eureka/lib/readECF.py:101]: In all the comment cases, I was afraid of deleting something until someone else verified my changes worked. (: Will delete now!

Comment (taylorbell57) [reply eureka/lib/readECF.py:101]: The nice thing is the edits are at least now saved in GitHub, so you can delete them and if ever you need them again you can try to search back through your commit history.

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/fitters.py:409]: This for loop looks to be identical to the for loops used in other fitter functions.  If so, it should called as a function instead of repeating the code.

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/fitters.py:462]: Let's move all of the log likelihood functions out of `fitters.py` and into a separate file

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/fitters.py:524]: Update file name

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/fitters.py:567]: Update file name

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/fitters.py:685]: Move `computeRMS` and `rmsplot` out of `fitters.py`

File: eureka/S5_lightcurve_fitting/lightcurve.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/lightcurve.py:146]: Might be time to update the name of this function from `plot` to `plotLC`, or something similarly descriptive.

File: eureka/S5_lightcurve_fitting/lightcurve.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/lightcurve.py:186]: Update file name

File: demos/S2/S2_nircam_wfss_template.ecf
Comment (taylorbell57) [reply demos/S2/S2_nircam_wfss_template.ecf:36]: I personally favour adding template files to the GitHub and keeping a local copy with personal edits. I'll revert this back to the genericized folders for now

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:180]: Now addressed in s5_tjbEdits

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:298]: Ah yes, I made this same mistake throughout. Now addressed in s5_tjbEdits

File: eureka/S5_lightcurve_fitting/lightcurve.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/lightcurve.py:146]: Well the function is already a class method, so it gets used like `lc = LightCurve(...); lc.plot(...)` which seems clear enough to me. We can still rename it if you'd like though

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:126]: All such comments are now addressed in s5_tjbEdits

File: eureka/S5_lightcurve_fitting/lightcurve.py
Comment (kevin218) [reply eureka/S5_lightcurve_fitting/lightcurve.py:146]: Good point!

File: demos/S2/S2_nircam_wfss_template.ecf
Comment (kevin218) [reply demos/S2/S2_nircam_wfss_template.ecf:36]: Agreed

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:409]: Now addressed in 25_tjbEdits

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:462]: Now addressed in s5_tjbEdits (moved to a new likelihood.py)

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:282]: Do you mean we should replace this hardcoded array with the diagonal of the **covariance** matrix that the scipy.optimize.leastsq function returns (probably multiplied by the variance of the residuals given that function's docstring [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.leastsq.html))?

Comment (kevin218) [reply eureka/S5_lightcurve_fitting/fitters.py:282]: Haha, oops.. yes, I meant covariance matrix.

Comment (taylorbell57) [reply eureka/S5_lightcurve_fitting/fitters.py:282]: Alright, this is now implemented in s5_tjbEdits too

**Comment by taylorbell57** [comment]
Alright, all of my comments and Kevin's are addressed (either here or in s5_tjbEdits which can be merged into here now). I'll try running S4->S5 with at least one instrument and I'll see what if anything breaks. Once that's functional we can merge this into main

**Comment by taylorbell57** [comment]
Alright, Megan and I have both run S4->S5 with different instruments and can confirm that S5 is functioning as we would expect it to at this moment. There is still significant work to do: implementing polynomials in time, implementing any detector models (especially a ramp model for MIRI), allowing different priors, etc. But for now we have something that seems well integrated with S4 and at least has the basic ability to fit a transit or eclipse. I vote we merge this into main now unless there are objections

File: eureka/S5_lightcurve_fitting/fitters.py
Comment (kevin218) [inline eureka/S5_lightcurve_fitting/fitters.py:261]: Optional

**Review by kevin218 — COMMENTED** [review]
For the next pull request, I suggest we convert all of the print statements to log statements so that the outputs are stored in a file.

**Comment by kevin218** [comment]
I got the polynomial fitting to work.  The bug was not obvious and required changes to the `PolynomialModel` class.  
I updated lsq.py to use `op.least_squares()` because `op.leastsq` was deprecated. I didn't go through the effort of converting the Jacobian matrix to the covariance matrix, so that's still a ToDo item.
I also fixed the 3 remaining issues I had from the last review.

**Review by kevin218 — APPROVED** [review]

**Pull request merged**
